official/resnet/ctl/ctl_imagenet_main.py

```diff
@@ -0,0 +1,295 @@
+from official.utils.misc import model_helpers
+from official.resnet.ctl import ctl_common
+
+
+# Imagenet training and test data sets.
+APPROX_IMAGENET_TRAINING_IMAGES = 1280000  # Approximate number of images.
+IMAGENET_VALIDATION_IMAGES = 50000  # Number of images.
+
+DEFAULT_IMAGE_SIZE = 224
+NUM_CHANNELS = 3
+NUM_CLASSES = 1001
+
+
+def parse_record_keras(raw_record, is_training, dtype):
+  """Adjust the shape of label."""
+  image, label = imagenet_main.parse_record(raw_record, is_training, dtype)
+
+  # Subtract one so that labels are in [0, 1000), and cast to float32 for
+  # Keras model.
+  label = tf.cast(tf.cast(tf.reshape(label, shape=[1]), dtype=tf.int32) - 1,
+                  dtype=tf.float32)
+  return image, label
+
+# Learning rate schedule
+_LR_SCHEDULE = [    # (multiplier, epoch to start) tuples
+    (1.0, 5), (0.1, 30), (0.01, 60), (0.001, 80)
+]
+_BASE_LEARNING_RATE = 0.4
+
+
+def compute_learning_rate(lr_epoch):
+  """Learning rate for each step."""
+  warmup_lr_multiplier, warmup_end_epoch = _LR_SCHEDULE[0]
+  if lr_epoch < warmup_end_epoch:
+    # Learning rate increases linearly per step.
+    return (_BASE_LEARNING_RATE * warmup_lr_multiplier *
+            lr_epoch / warmup_end_epoch)
+  for mult, start_epoch in _LR_SCHEDULE:
+    if lr_epoch >= start_epoch:
+      learning_rate = _BASE_LEARNING_RATE * mult
+    else:
+      break
+  return learning_rate
+
+
+def run(flags_obj):
+  """Run ResNet ImageNet training and eval loop using custom training loops.
+
+  Args:
+    flags_obj: An object containing parsed flag values.
+
+  Raises:
+    ValueError: If fp16 is passed as it is not currently supported.
+
+  Returns:
+    Dictionary of training and eval stats.
+  """
+  dtype = flags_core.get_tf_dtype(flags_obj)
+
+  # TODO(anj-s): Set data_format without using Keras.
+  data_format = flags_obj.data_format
+  if data_format is None:
+    data_format = ('channels_first'
+                   if tf.test.is_built_with_cuda() else 'channels_last')
+  tf.keras.backend.set_image_data_format(data_format)
+
+  strategy = distribution_utils.get_distribution_strategy(
+      distribution_strategy=flags_obj.distribution_strategy,
+      num_gpus=flags_obj.num_gpus,
+      num_workers=distribution_utils.configure_cluster(),
+      all_reduce_alg=flags_obj.all_reduce_alg,
+      num_packs=flags_obj.num_packs)
+
+
+  # pylint: disable=protected-access
+  if flags_obj.use_synthetic_data:
+    input_fn = keras_common.get_synth_input_fn(
+        height=imagenet_main.DEFAULT_IMAGE_SIZE,
+        width=imagenet_main.DEFAULT_IMAGE_SIZE,
+        num_channels=imagenet_main.NUM_CHANNELS,
+        num_classes=imagenet_main.NUM_CLASSES,
+        dtype=dtype,
+        drop_remainder=True)
+  else:
+    input_fn = imagenet_main.input_fn
+
+  train_ds = input_fn(
+      is_training=True,
+      data_dir=flags_obj.data_dir,
+      batch_size=flags_obj.batch_size,
+      num_epochs=flags_obj.train_epochs,
```

Comment:
Since you are creating a new iterator every epoch, you don't need the num_epochs.

However, this does bring up the shuffle-repeat issue. Since you're going through the dataset from the beginning every epoch, it would be going through the data in the same order every epoch. Does that still converge?

Reply:
This did converge because there was a shuffle but no repeat. Wouldn't that ensure that the input is randomized every epoch? Or did you mean that if I repeat() with shuffle I would get the same set of random input every epoch?

Reply:
so there is an issue with shuffle currently that without a repeat, it doesn't really shuffle every time you create an iterator on the same dataset. rohan is fixing it but i dont think it is yet done. 

i can explain in person tomorrow, it's a bit complex to write down =)